GoogleUtilities/FileDataStorage/GULHeartbeatDateStorage.m

```diff
@@ -0,0 +1,146 @@
+  NSError *fileCoordinatorError = nil;
+  [fileCoordinator
+      coordinateWritingItemAtURL:directoryPathURL
+                         options:0
+                           error:&fileCoordinatorError
+                      byAccessor:^(NSURL *writingDirectoryURL) {
+                        NSError *error;
+                        if (![writingDirectoryURL checkResourceIsReachableAndReturnError:&error]) {
+                          // If fail creating the Application Support directory, log warning.
+                          NSError *error;
+                          GULLoggerService kGULHeartbeatDateStorage = @"GULHeartbeatDateStorage";
+                          if (![[NSFileManager defaultManager]
+                                         createDirectoryAtURL:writingDirectoryURL
+                                  withIntermediateDirectories:YES
+                                                   attributes:nil
+                                                        error:&error]) {
+                            GULLogWarning(kGULHeartbeatDateStorage, YES, @"I-COR100001",
+                                          @"Unable to create internal state storage: %@", error);
+                          }
+                        }
+                      }];
+}
+
+- (nullable NSMutableDictionary *)heartbeatDictionary {
+  NSError *fileCoordinatorError = nil;
+  __block NSMutableDictionary *dict;
+  [self.fileCoordinator
+      coordinateReadingItemAtURL:self.fileURL
+                         options:0
+                           error:&fileCoordinatorError
+                      byAccessor:^(NSURL *readingFileUrl) {
+                        NSError *error;
+                        NSData *objectData = [NSData dataWithContentsOfURL:readingFileUrl
+                                                                   options:0
+                                                                     error:&error];
+                        if (objectData == nil || error != nil) {
+                          dict = [NSMutableDictionary dictionary];
+                        } else {
+                          dict = [GULSecureCoding unarchivedObjectOfClass:NSObject.class
+                                                                 fromData:objectData
+                                                                    error:&error];
+                          if (dict == nil || error != nil) {
+                            dict = [NSMutableDictionary dictionary];
+                          }
+                        }
+                      }];
+  return dict;
+}
+
+- (nullable NSDate *)heartbeatDateForTag:(NSString *)tag {
+  NSMutableDictionary *dictionary = [self heartbeatDictionary];
+  return dictionary[tag];
+}
+
+- (BOOL)setHearbeatDate:(NSDate *)date forTag:(NSString *)tag {
+  NSMutableDictionary *dictionary = [self heartbeatDictionary];
+  dictionary[tag] = date;
```

Comment:
I am afraid we may have a race condition here which can lead to lost updates:
- two thread read dict
- both modify their own dict copy
- the one which writes last overwrites changes from the one which writes first.

Looks like [this method](https://developer.apple.com/documentation/foundation/nsfilecoordinator/1413385-coordinatereadingitematurl?language=objc) can help us to solve it. So we will need to have a single coordinated operation for reading, modification and writing.